### PR TITLE
SapMachine #1449: Store vitals samples for max/min values

### LINKED ISSUE #1449: Store vitals samples for max/min values
In addition to the short and long time tables, store a sample when specific values are at a maximum or minimum. This not only contains the information about the max/min value, but has the whole set of values at that time. So you can see for example the memory usage of the different types when the free memory was lowest or the committed memory the highest. For each value we store decide if we are interested in the max value, the min value or neither.

### PR DESCRIPTION
**Pull request by schmelter-sap**
This stores the extreme samples for specific values. For most values the sample at the maximum is stored, but for some the minimum sample is stored (e.g. free memory). The samples are included at the end of the report.

Some additional changes were made:
- Fixed small memory leak in the ColumnWidth class.
- We don't use the temporary pre-allocated buffer when dumping the report but instead use the supplied outputStream directly. Blocking when writing to the stream should at most hinder updates to the vitals for the time we are blocked. 
- Fixed wrong string comparison when handling categories in legends.

fixes #1449

### CODE REVIEW

**Comment by SapMachine** [comment]
Hello @schmelter-sap, I'm sorry, this pull request doesn't meet the expectations. The pull request description has an invalid format. Please have a look at https://github.com/SAP/SapMachine/wiki/Formal-Requirements-of-Pull-Requests .

**Comment by RealCLanger** [comment]
Hi @schmelter-sap,

you need to create an Issue here in GitHub first, with a common description of what you intend to fix. (Like a JBS bug). Then, reference it in the PR description with `fixes #<issue number>` in its last line. Then your PR would fulfil the formal requirements.

**Comment by RealCLanger** [comment]
retest this please

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Review by MBaesken — APPROVED** [review]

File: src/hotspot/share/vitals/vitals.cpp [hunk 1]
```diff
@@ -829,64 +837,117 @@ class SampleTables: public CHeapObj<mtInternal> {
...
+      } else if (_extremum_samples.is_empty()) {
+        // We already have a last sample and this is the second sample we see.
+        // We can initialize the two tables now to store the last sample and this sample
+        // for all extremas.
+        for (int i = 0; i < Sample::num_values(); ++i) {
+          _last_extremum_samples.add_sample(last_sample);
+          _extremum_samples.add_sample(sample);
+        }
+      }
+      else {
```
Comment (RealCLanger) [inline src/hotspot/share/vitals/vitals.cpp:859]: `else` should be in line 858 to match formatting in file

File: src/hotspot/share/vitals/vitals_internals.hpp [hunk 2]
```diff
@@ -93,9 +100,11 @@ namespace sapmachine_vitals {
     const char* header() const        { return _header; }
     const char* name() const          { return _name; }
     const char* description() const   { return _description; }
+    Extremum    extremum() const      { return _extremum; }
 
     void print_value(outputStream* os, value_t value, value_t last_value,
-        int last_value_age, int min_width, const print_info_t* pi) const;
+        int last_value_age, int min_width, const print_info_t* pi,
```
Comment (RealCLanger) [inline src/hotspot/share/vitals/vitals_internals.hpp:106]: you can keep this in one line

**Review by RealCLanger — APPROVED** [review]
Some minor formatting comments

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Pull request merged**